### PR TITLE
add API layout strategy

### PR DESCRIPTION
**Pull request by thomas-maschler**
**Related Issue(s):**

- https://github.com/stac-utils/pystac-client/issues/241

**Description:**
This PR implements the API Layout Strategy following the STAC API Specifications.
Unlike the Best Practice Layout Strategy, this strategy is not based on file paths. Instead, it uses HTTP endpoints. This requires refactoring of Catalog's `add_child`, `add_item`, and  `update_href` methods not to remove the last path element. Instead, I moved this step into the `HrefLayoutStrategy.get_href` method and made it conditional based on whether the path is a file or not.

Together with https://github.com/stac-utils/pystac/pull/1295, this PR is part of setting the foundation for enabling API transactions using `pystac-client`.

**PR Checklist:**

- [x] `pre-commit` hooks pass locally
- [x] Tests pass (run `scripts/test`)
- [x] Documentation has been updated to reflect changes, if applicable
- [x] This PR maintains or improves overall codebase code coverage.
- [x] Changes are added to the [CHANGELOG](https://github.com/stac-utils/pystac/blob/main/CHANGELOG.md). See [the docs](https://pystac.readthedocs.io/en/latest/contributing.html#changelog) for information about adding to the changelog.

### CODE REVIEW

File: pystac/utils.py [hunk 1]
```diff
@@ -559,3 +559,8 @@ def get_required(option: T | None, obj: str | Any, prop: str) -> T:
     if option is None:
         raise RequiredPropertyMissing(obj, prop)
     return option
+
+
+def is_file_path(href: str) -> bool:
```
Comment (jsignell) [inline pystac/utils.py:564]: It's pretty self-explanatory, but a docstring here would be great.

**Review by jsignell — APPROVED** [review]
Sorry for the slow response. This looks really good to me. I was kind of surprised that this didn't exist yet in pystac.

**Comment by thomas-maschler** [comment]
Thanks for the review. I added a docstring

**Comment by thomas-maschler** [comment]
@jsignell I had to resolve a conflict in the change log. Can you resubmit?

**Pull request merged**